docs/hugo/content/design/ADR-2023-01-Conversion-Augmentation.md

```diff
@@ -0,0 +1,135 @@
+---
+title: '2022-12: Conversion Augmentation'
+---
+
+## Context
+
+The Kubernetes ecosystem has [strong requirements](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html) around backward and forward compatibility of resource versions:
+
+> In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information. Thus, any change we make to our API must be compatible with whatever we supported in v1, and also need to make sure anything we add in v2 is supported in v1.
+
+To achieve this, we are [code generating conversions](https://azure.github.io/azure-service-operator/design/versioning/) between versions of our resources.
+
+However, we've always known that we'd need the capability to augment these conversions with some hand-written code, if only because we would quickly hit a point of diminishing returns if we try to handle everything with the generator. 
+
+There are three different ways we can implement this:
+
+* Generic interfaces
+* Code-gen interfaces
+* Configuration
+
+### Generic interface implementation
+
+In our `genruntime` package, we'll define a new generic interface to capture the semantics of conversion augmentation:
+
+```go
+type Assignable[T conversion.Convertible] interface {
+	AssignPropertiesTo(dst T) error
+	AssignPropertiesFrom(src T) error
+}
+```
+Conversions always need to be bi-directional, so we force both `AssignPropertiesTo` and `AssignPropertiesFrom` to be implemented together.
+
+We then update the code generation for the `AssignProperties_To_*` and `AssignProperties_From_*` methods to look for this interface and call it if it's present. The type used for the type parameter will be the "upstream" type in our conversion chain.
+
+For example, for the `v1beta20210101` version of `BatchAccount`, the existing method `AssignProperties_From_BatchAccount` will updated with the following additional code:
+
+``` go
+var sourceAsAny any = source
+if src, ok := sourceAsAny.(genruntime.Assignable[*v20210101s.BatchAccount]); ok {
+    err := src.AssignFrom(source)
+    if err != nil {
+        return errors.Wrap(
+            err, 
+            "calling custom AssignFrom() for conversion from v20210101s.BatchAccount")
+    }
+}
+```
+
+**Pro**: Minimizes the amount of code we need to generate because we only need an extra clause in the existing conversion methods.
```

Comment:
Would it be in conversion methods for every type? Or just for top level resource type?

I assume the proposal here is just for the top level resource type because as I understand it, that's the type which implements `conversion.Convertible`

Reply:
I've updated the ADR to explicitly call this out, and included details from our earlier discussion indicating that implementing it for all resources and objects makes things easier.